Step3 구현

joont92 opened the pull request:
많은 부분을 변경하였습니다.

1. Service
Service에서 Repository까지 통합테스트 하는식으로 작성하였습니다.

2. Acceptance Test
이 부분이 정말 어려웠습니다.
(처음에는 테스트마다 전부 픽스쳐를 세팅하고, 끝나면 롤백되는 형식으로 구현하다가 다 제거했었습니다.. )
매번 픽스쳐 세팅 + 롤백 으로 작성하니 테스트가 고립되기는 하는데, 뭔가 굉장히 복잡해지는 느낌이었고,
공통 데이터(import.sql)을 사용하니 테스트가 고립되지 않는 느낌이었고, 검증하는 부분도 상당히 제한적이게 된 느낌이었습니다..
(결과적으로 지금은 2개를 반쯤 섞어서? 하고 있습니다..)

부족하다보니 시간이 많이걸리네요
리뷰 부탁드립니다! 감사합니다!

joont92 committed:
feat : question api 작성

- ApiQuestionAcceptanceTest 작성
- ApiQuestionController 작성
- AcceptanceTest 공통 메서드 추가

joont92 committed:
feat, refactor : Question 기능 구현

- QnAServiceTest, QuestionTest 작성
	- 기존 mock 제거(h2 사용)
- QnAService, Question 구현

joont92 committed:
test : QnA AcceptanceTest fixture 변경

- 질문 생성시 오류 변경
- ApiQnAAcceptanceTest fixture 수정

joont92 committed:
feat : 기능 1차 완료

- QnaService, QnaServiceTest 완료
- Answer, Question 구현 및 단위테스트 구현

joont92 committed:
step 3 완료

joont92 committed:
fix : missing delete

phs1116 commented on src/test/java/nextstep/service/QnAServiceTest.java:
```diff
@@ -1,104 +1,113 @@
+    private Question fixtureQuestion;
+    private Answer fixtureAnswer;
+
+    @Autowired private QnAService qnaService;
```
음 QnAServiceTest는 유닛테스트입니다. 이렇게 테스트하시면 QnAService 자체 내부 로직에 대한 검증이 이루어지기 어렵습니다. 
(현재는 Repository에만 의존을 가지고 있지만 그 외에 다른 인프라스트럭쳐 영역이나 다른 도메인에도 의존을 가지고 있으면 더더욱이요.)
현재는 서비스가 가벼워서 그닥 체감을 못하시겠지만 TDD를 할 경우  스프링 컨테이너를 띄우는 시간 때문에 개발-테스트에 대한 간극도 많이 커질 수도 있어요.

유닛 테스트에서외부 의존을 가지고 있는 부분들을 Mock으로 처리하는 이유가 순수하게 이 서비스의 로직에만 집중하기 위해서입니다. 또한 Mock으로 처리하지 않을 경우에 해당 기능에 추가 기능이 있을 경우 이 서비스가 의존해야 하는 빈을 모두 구현해야한다는 어려움도 있구요.
Repository에 대한 빈을 직접생성하고 실제 빈을 주입받는 것 보다는 목으로 처리하셔서 QnAService 자체에 집중하시는게 좋을 것 같아요!
fixture도 실제 DB에 의존을 가지고 하는것 보다는 Mockito같은 유틸을 써서 사용해보시는게 좋을 것 같네요.

지금 하신것처럼 SpringRunner를 외부 의존까지 모두 가져와서 서비스의 스펙을 테스트 하는 부분은 인수테스트에서 검증되니 거기서 테스트하시면 될 것 같습니다.

joont92 replied:
말씀하신부분 적용해봤습니다. 피드백 정말 감사드립니다!

phs1116 commented on src/test/java/support/test/AcceptanceTest.java:
```diff
@@ -39,4 +42,53 @@ protected User defaultUser() {
+        return response;
+    }
+
+    protected <P, T> ResponseEntity<T> updateResource(String url, P bodyPayload, Class<T> responseType, User loginUser) {
```
이러한 API 테스트 메소드들을 HtmlFormDataBuilder처럼 하나의 유틸 클래스로 만들어보면 어떨까요?

joont92 replied:
RestApiTestCaller를 만들어봤습니다

phs1116 commented on src/main/java/nextstep/domain/QuestionBody.java:
```diff
@@ -0,0 +1,52 @@
+import javax.validation.constraints.Size;
+
+@Embeddable
+public class QuestionBody {
```
QuestionBody 분리 👍

phs1116 commented on src/test/java/nextstep/web/ApiQnAAcceptanceTest.java:
```diff
@@ -0,0 +1,183 @@
+        // then
+        String location = createResponse.getHeaders().getLocation().getPath();
+        Question question = getResource(location, Question.class, givenUser).getBody();
+        softly.assertThat(question.getQuestionBody()).isEqualTo(payload);
```
body대한 검증만 있으면 될까요?
보통 API스펙에서는 HttpStatus도 같이 작성하는 경우가 많습니다.
클라이언트 입장에서 HttpStatus도 검증되는게 더 좋지 않을까요?

joont92 replied:
상단에 createResource 메서드에 201을 검증하는 코드가 들어있어요!

phs1116 commented on src/test/java/nextstep/web/ApiQnAAcceptanceTest.java:
```diff
@@ -0,0 +1,183 @@
+        Question givenQuestion = generateQuestion(givenUser);
+
+        // when
+        deleteResource(String.format("/api/questions/%d", givenQuestion.getId()), givenUser);
```
해당 테스트에 대한 검증이 없는 것 같습니다.

joont92 replied:
추가하였습니다!

phs1116 reviewed (changes requested):
안녕하세요. 리뷰어 박현식입니다.
QuestionBody 클래스 분리 👍 
테스트 쪽은 짜시느라 고민 많이 하신 흔적이 보여요!
테스트 코드 쪽에 몇가지 피드백 드렸는데요, 
확인하시구 궁금하신점 있으시면 연락주세요. :)

joont92 commented on src/main/java/nextstep/service/QnAService.java:
```diff
@@ -24,54 +24,72 @@
+
+    @Deprecated
+    @Transactional(readOnly = true)
+    public Question findByOwner(User loginUser, Long id) {
```
제거했습니다

joont92 commented on src/main/java/nextstep/web/ApiQnAController.java:
```diff
@@ -0,0 +1,81 @@
+@Controller
+@RequestMapping("/api/questions")
+public class ApiQnAController {
+    private static final String urlPrefix = "/api/";
```
수정했습니다!

joont92 committed:
fix : 서비스 테스트 변경

- 실제 repository -> mock으로 변경

joont92 commented on src/main/java/nextstep/service/QnAService.java:
```diff
@@ -24,54 +24,72 @@
-    public Answer deleteAnswer(User loginUser, long id) {
-        // TODO 답변 삭제 기능 구현 
+    public Answer deleteAnswer(User writer, Long id) {
+        // TODO 답변 삭제 기능 구현
```
구현하였습니다!

joont92 committed:
feat, refactor, fix : step3 피드백 반영

- 도메인에서 사용하지 않는 메서드 삭제
- 답변 삭제 기능 구현
- RestApiTestCaller 추가
- AcceptanceTest 전반적 검토 및 수정

joont92 commented:
steo3을 진행하면서 많은 부분이 해소된 것 같습니다
도움을 주셔서 정말 감사합니다!

리뷰 부탁드립니다!

phs1116 reviewed (approved):
안녕하세요. 리뷰어 박현식입니다. :)
피드백 적극 반영해주셔서 감사합니다. 
RestApiTestCaller 분리 👍 
유닛테스트도 전체적으로 잘 짜신 것 같아요.
3단계 머지하겠습니다.
3단게 구현하시느라 수고 많으셨습니다. :)
